[Tradfri] Support for new RGB bulbs #4251

hreichert opened the pull request:
Issue: #4251

Add support for the new TRADFRI full color lamps.
* New ThingType _0210_ / Extended Color Light with additional channel _color_
* Extended _TradfriLightHandler_
    * Handle color setting and retrieving
    * Changed color temperature calculation
* New class TradfriColor for colorspace conversions
* Extended _TradfriDiscoveryService_ for automatic discovery of color lights
* Tests for the above
* Documentation in README.md

I think it's ready for review.

hreichert commented on extensions/binding/org.eclipse.smarthome.binding.tradfri/src/main/java/org/eclipse/smarthome/binding/tradfri/handler/TradfriLightHandler.java:
```diff
@@ -243,17 +267,51 @@ private void handleColorTemperatureCommand(Command command) {
+            setColor((HSBType) command);
+            scheduler.schedule(() -> {
+                setBrightness(((HSBType) command).getBrightness());
+            }, 1500, TimeUnit.MILLISECONDS);
```
Maybe it's better to calculate the delay based on the dimmer transition time?

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.tradfri/ESH-INF/thing/thing-types.xml:
```diff
@@ -71,6 +71,28 @@
+        <description>A dimmable light that supports full colors and color temperature settings.</description>
+
+        <channels>
+            <channel id="brightness" typeId="brightness"/>
```
The color channel already accepts all the same (PercentType) commands as the brightness channel, so a thing that has the color channel, should not have the brightness channel as well as this is redundant.
See e.g. also the hue + LIFX bindings where it is the same.

hreichert replied:
Ok, I can remove the channel. But I think that we loose functionality:
Imagine a group gWohnzimmer_Brightness for the living room.
In this group we have
* Something_Wohnzimmer_Brightness (something something)
* Tradfri_Wohnzimmer1_Brightness (dimmable bulb)
* Tradfri_Wohnzimmer2_Brightness (dimmable color temperature bulb)
* Tradfri_Wohnzimmer3_Brightness (full color bulb)

Further we have a `Slider item=gWohnzimmer_Brightness` in the sitemap.
Slide the slider and see how all lamps, also the full color bulb, change their brightness.
Without the brightness channel, one would have to use some rules, I think?
Or do I have a knot in my brain?

htreu replied:
Iirc when sending only the brightness value (0 - 100) to a color item it will be interpreted as brightness only. So in your example all bulbbs should change their brightness.

hreichert replied:
Ah well I think you're right.
Will verify this and update my PR accordingly.

sjka replied:
You remember correctly! The [ColorItem](https://github.com/eclipse/smarthome/blob/74a486dfbf8201a2c48f021fa2ef4353a5ba2ca8/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/items/ColorItem.java#L83) takes care of this: If it gets updated with a percent type, it keeps everything else from the previous state and only adjusts the brightness.

hreichert replied:
I tried this, but got unexpected results.
My steps:
* Breakpoints on `TradfriLightHandler.handleCommand` and `ColorItem.setState`
* Item: tradfri_0200_gwa0cc2b6d64bb_65551_color
* Set a color with BasicUI -> handleCommand receives HSBType -> good
* Send "100" to the same item -> handleCommand receives PercentType -> unexpected?
* AFTER this, the breakpoint on `ColorItem.setState` kicks in
My current code has already a special handling for PercentType on the color channel, so the functionality is OK. But the ColorItem.setState doesn't help here, I think.
What's your opinion about this?

cweitkamp commented on extensions/binding/org.eclipse.smarthome.binding.tradfri/ESH-INF/thing/thing-types.xml:
```diff
@@ -71,6 +71,28 @@
         </config-description>
     </thing-type>
 
+    <thing-type id="0200">
```
In case of a color lamp with adjustable color temperature and brightness the corresponding ZigBee ID should be 210 instead of 200.

hreichert replied:
The _real_ device ID of this lamps is 200 (0x0200), please see #4251 (confirmed on hardware).
Do we take the _real_ device ID, or the device ID that we imagine that is correct?
The whole thingType naming based on technical device IDs, that the IKEA gateway do not even expose, is anyway a bit.. hmm

kaikreuzer replied:
When creating it this way I was hoping that IKEA would stay close to the specification, which would be the best way to figure out the capabilities of the device. This is what we ended up with in the hue binding after many 3rd party devices were supported by it. I'd guess that sooner or later IKEA will come to the same conclusion and our approach will prove to be the right choice ☺️

cweitkamp replied:
Sry didn't look into #4251. You already complained your choice. And this is another break with the ZigBee protocol caused by IKEA developers.

To be compliant I would suggest to use the same thing-type definition as we use in the Hue and other bindings. In this case it should be the device ID of what we think the bulb is capable of -> 210. If not, we should remove the color_temperature channel as well (where later one is IMHO the worse way).

btw. I didn't mentioned it before: You did a great job so far.

hreichert replied:
My personal feeling is also that 210 is the right choice, and not 200. But as the _real_ device ID was 210, I renamed it so stick with the previous naming scheme.
I'm happy to re-rename it to 210, cause this better reflects the capabilities of the Thing.

P.S: After re-reading my previous comment, it sounded a bit offensive. That was not intentional.

cweitkamp replied:
> But as the real device ID was 210, ...

You mean 200, I guess. ;-) Okay, we give it a try.

> ... it sounded a bit offensive.

Oh no, everything was alright. I am fine with it.

hreichert replied:
Yes, 200... All these numbers!
I changed it now to *210* in the binding and the documentation. PR update ongoing with the other changes.

kaikreuzer replied:
Hm, my comment above actually meant that we SHOULD go for the real Zigbee type, because sooner or later other devices will appear with the 210 type - and maybe they then need a different logic to be supported (e.g. hardware instead of software color temp setting). So I was actually fine with the "200" choice.

>  If not, we should remove the color_temperature channel as well

@cweitkamp The other option would be that we think about doing "software calculation" on those (hue) as well - do you think that would be possible?

kaikreuzer replied:
But I don't have a too strong opinion on 200 vs. 210 - if you both now agreed that 210 is the better choice, so be it. We can still revisit this once we see more devices that bring us more clarity on how the gateway handles them.

hreichert replied:
Well, @cweitkamp has also good arguments.
I'm absolutely against removing the color_temperature channel from the new thing. See my discussion above; but replace "brightness" with "color temperature". And this can not be handled with the color channel, at least not without rules or something.

hreichert commented:
Updated with following changes:
* Renamed ThingType 200 to 210
* Removed "brightness" channel from ThingType 210
* Improved color calculations
* Documentation

kaikreuzer commented on extensions/binding/org.eclipse.smarthome.binding.tradfri/README.md:
```diff
@@ -12,14 +12,15 @@ The thing type ids are defined according to the lighting devices defined for Zig
 |-------------|:------:|:----------:|:-----:|:-----------------:|   
 |  0100       |    X   |     X      |       |                   |
 |  0220       |    X   |     X      |       |          X        |
+|  0210       |    X   |            |   X   |          X        |
```
I am noticing that this table was already incorrect - neither 0220 nor 0210 should have an X for On/Off...

hreichert replied:
Very good point, as there is no "onoff" channel. Should I open a new issue for this? I want to extend the documentation anyway with more examples.

kaikreuzer replied:
You could simply remove the column as part of this PR - or we first merge and you can come up with a new PR that enhances the documentation in general.

hreichert replied:
I vote for option 2 - new PR when the enhancements are done.

kaikreuzer replied:
ok!
